posts/2021-05-12-paying-homage-to-the-compall-michael-representation.adoc

```diff
@@ -0,0 +1,394 @@
+
+[source,scala]
+----
+object Units {
+  sealed trait Distance extends Any
+  final case class Kilometres private[Units] (value: Double) extends AnyVal with Distance
+  final case class Miles private[Units] (value: Double) extends AnyVal with Distance
+
+  ...
+
+  def toKm(distance: Distance): Kilometres = distance match {
+    case miles: Miles => Kilometres(miles.value * 1.6)
+    case kilometres: Kilometres => kilometres
+  }
+}
+----
+
+This code is perfectly fine, but if you had chosen the `AnyVal` route to avoid allocations, then you would be disappointed to discover that now with the supertype-subtype relationship you will once again have allocations when you pass a value of `Kilometres` or `Miles` to the function `toKm`.
+
+[source,scala]
+----
+import Units._
+
+...
+
+class Rocket(booster: Booster) {
+  private var distance: Kilometres = ZeroKm
+
+  def launch(): Unit = {
+    val launchBoost: Kilometres = toKm(booster.provideLaunchBoost()) // Allocation of Miles object
+    distance = add(distance, launchBoost)
+  }
+
+  def distanceTravelled: Kilometres = distance
+}
+----
+
+The various limitations of Value-classes are extensively discussed on the https://docs.scala-lang.org/overviews/core/value-classes.html#limitations[Scala Lang page on Value Classes] and on the https://docs.scala-lang.org/sips/value-classes.html[SIP-15 page] there is recognition that improvements can be made for certain use-cases (e.g. numerical computing).
+
+== What it looks like with the "Compall-Michael" pattern
+In contrast to the preceding well-known and widely-used techniques seen thus far in the post, the approach that I am calling here the "Compall-Michael" pattern is almost certainly less well known (even though in some respects it already has a very long heritage in Scala). This technique makes use of Scala's ability to emulate the ML module system, which I first heard Martin Odersky discuss in his flatMap Oslo talk in 2014 (https://www.slideshare.net/Odersky/flatmap-oslo-presentation-slides[click here for the slides]) and which is very well explained in this blog post from that same year entitled https://pellucidanalytics.tumblr.com/post/94532532890/scalas-modular-roots-earlier-this-year-martin["Scala's Modular Roots" by Dan James].
+
+For an in-depth explanation of how the emulation of ML modules in Scala can be achieved, I invite you to read the blog post linked above (and if you want a crash course on ML modules then https://jozefg.bitbucket.io/posts/2015-01-08-modules.html[this blog post is a good quick read]). Here I will just say that the main aspects that interest us are emulating ML _signatures_ using `sealed trait`s and emulating ML _structures_ by implementing the signature with `new { ... }` and assigning it to a `val`. (In our use case we will not need to emulate ML _functors_ which is the third important aspect of the ML module system).
```

Comment:
the rendering of "using `sealed trait`s and emulating ML _structures_ by implementing the signature with `new { ... }`" seems to be messed up in the preview. But I don't know why 🤔.

Reply:
Ok I think I see the problem. Hopefully it should be fixed with the change I have made.